File: packages/stera_recorder/android/src/main/kotlin/open/fpvlabs/stera/ar_recorder/session/FrameProcessingPipeline.kt
package open.fpvlabs.stera.ar_recorder.session

import com.google.ar.core.Camera
import com.google.ar.core.Frame
import com.google.ar.core.TrackingState

class FrameProcessingPipeline {
    data class TrackingTransitionResult(
        val trackingPausedAtNs: Long,
        val trackingPauseCount: Int,
        val trackingPauseTotalOffsetNs: Long,
        val wasTrackingInPreviousFrame: Boolean,
        val depthCadenceResetCount: Int
    )

    fun evaluateTrackingTransition(
        timestampNs: Long,
        camera: Camera,
        trackingPausedAtNs: Long,
        trackingPauseCount: Int,
        trackingPauseTotalOffsetNs: Long,
        wasTrackingInPreviousFrame: Boolean,
        depthCadenceResetCount: Int,
        onTrackingPauseOffset: (Long) -> Unit
    ): TrackingTransitionResult {
        var pausedAt = trackingPausedAtNs
        var pauseCount = trackingPauseCount
        var pauseOffsetNs = trackingPauseTotalOffsetNs
        var cadenceResetCount = depthCadenceResetCount
        val isTracking = camera.trackingState == TrackingState.TRACKING
        if (!isTracking) {
            if (pausedAt == Long.MIN_VALUE) {
                pausedAt = timestampNs
                pauseCount++
            }
        } else if (pausedAt != Long.MIN_VALUE) {
            val offsetNs = (timestampNs - pausedAt).coerceAtLeast(0L)
            pauseOffsetNs += offsetNs
            onTrackingPauseOffset(offsetNs)
            pausedAt = Long.MIN_VALUE
        }
        if (isTracking && !wasTrackingInPreviousFrame) {
            cadenceResetCount++
        }
        return TrackingTransitionResult(
            trackingPausedAtNs = pausedAt,
            trackingPauseCount = pauseCount,
            trackingPauseTotalOffsetNs = pauseOffsetNs,
            wasTrackingInPreviousFrame = isTracking,
            depthCadenceResetCount = cadenceResetCount
        )
    }

    fun shouldProcess(isPaused: Boolean, frame: Frame): Boolean {
        return !isPaused && frame.timestamp > 0L
    }
}

File: packages/stera_recorder/ios/stera_recorder/Sources/stera_recorder/session/FrameProcessingPipeline.swift
import Foundation
import ARKit

class FrameProcessingPipeline {

    struct TrackingTransitionResult {
        let trackingPausedAtNs: Int64
        let trackingPauseCount: Int
        let trackingPauseTotalOffsetNs: Int64
        let wasTrackingInPreviousFrame: Bool
        let depthCadenceResetCount: Int
    }

    func evaluateTrackingTransition(
        timestampNs: Int64,
        trackingState: ARCamera.TrackingState,
        trackingPausedAtNs: Int64,
        trackingPauseCount: Int,
        trackingPauseTotalOffsetNs: Int64,
        wasTrackingInPreviousFrame: Bool,
        depthCadenceResetCount: Int,
        onTrackingPauseOffset: (Int64) -> Void
    ) -> TrackingTransitionResult {
        var pausedAt = trackingPausedAtNs
        var pauseCount = trackingPauseCount
        var pauseOffsetNs = trackingPauseTotalOffsetNs
        var cadenceResetCount = depthCadenceResetCount

        let isTracking: Bool
        switch trackingState {
        case .normal:
            isTracking = true
        default:
            isTracking = false
        }

        if !isTracking {
            if pausedAt == Int64.min {
                pausedAt = timestampNs
                pauseCount += 1
            }
        } else if pausedAt != Int64.min {
            let offsetNs = max(timestampNs - pausedAt, 0)
            pauseOffsetNs += offsetNs
            onTrackingPauseOffset(offsetNs)
            pausedAt = Int64.min
        }

        if isTracking && !wasTrackingInPreviousFrame {
            cadenceResetCount += 1
        }

        return TrackingTransitionResult(
            trackingPausedAtNs: pausedAt,
            trackingPauseCount: pauseCount,
            trackingPauseTotalOffsetNs: pauseOffsetNs,
            wasTrackingInPreviousFrame: isTracking,
            depthCadenceResetCount: cadenceResetCount
        )
    }

    func shouldProcess(isPaused: Bool, frameTimestamp: TimeInterval) -> Bool {
        return !isPaused && frameTimestamp > 0
    }
}
